source/agora/test/Base.d

```diff
@@ -744,6 +753,31 @@ public APIManager makeTestNetwork (APIManager : TestAPIManager = TestAPIManager)
         return conf;
     }
 
+    // for disovery testing: full quorum, but only 1 node in the 'network' section
+    Config makeMinimalNetQuorum (size_t idx, NodeConfig self, NodeConfig[] node_confs)
+    {
+        auto prev_node = idx == 0 ? node_confs[$ - 1] : node_confs[idx - 1];
+
+        auto quorum_keys =
+            node_confs
+                .filter!(conf => conf.is_validator)
```

Comment:
If conf.is_validator is false, an exception may occur.

> Module tests failed: agora.test.NetworkDiscovery
object.Exception@submodules/localrest/source/geod24/LocalRest.d-mixin-716(770): "Request timed-out"
----------------
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 @safe void geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI.start() [0x10e7876da]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 @safe void agora.test.Base.TestAPIManager.start().__lambda1!(agora.test.Base.NodePair).__lambda1(agora.test.Base.NodePair) [0x10e7965e0]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 @safe std.typecons.Flag!("each").Flag std.algorithm.iteration.each!(agora.test.Base.TestAPIManager.start().__lambda1).each!(agora.test.Base.NodePair[]).each(ref agora.test.Base.NodePair[]) [0x10e796657]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 void agora.test.Base.TestAPIManager.start() [0x10e783b7f]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 void agora.test.NetworkDiscovery.__unittest_L69_C1() [0x10e7ab21a]
source/agora/node/main.d:40 void agora.test.NetworkDiscovery.__modtest() [0x10e4edf4a]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 int agora.test.Base.customModuleUnitTester().__foreachbody3(ref agora.test.Base.customModuleUnitTester().ModTest) [0x10e7837df]
/usr/local/opt/dmd/include/dlang/dmd/core/internal/array/equality.d:33 void std.parallelism.ParallelForeach!(agora.test.Base.customModuleUnitTester().ModTest[]).ParallelForeach.opApply(scope int delegate(ref agora.test.Base.customModuleUnitTester().ModTest)).doIt() [0x10e566203]

Reply:
if *none* of the node configs `is_validator` is true, it will fail because `map` will try to invoke `popFront` on the empty range. But this `makeMinimalNetQuorum` should only be called for this specific test-case where all the nodes are validators.

Reply:
Will it ? I would expect array on an empty range to return `null`.
Perhaps add an `assert` that `.length > 0` for clarity ?

Reply:
Ah it definitely does not throw. I wonder where I got this error from.

I'll add an assert.